controllers/kustomization_varsub.go

```diff
@@ -0,0 +1,90 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/drone/envsubst"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/kustomize/api/resource"
+	"sigs.k8s.io/yaml"
+
+	kustomizev1 "github.com/fluxcd/kustomize-controller/api/v1beta1"
+)
+
+// substituteVariables replaces the vars with their values in the specified resource.
+// If a resource is labeled or annotated with
+// 'kustomize.toolkit.fluxcd.io/substitute: disabled' the substitution is skipped.
+func substituteVariables(
+	ctx context.Context,
+	kubeClient client.Client,
+	kustomization kustomizev1.Kustomization,
+	res *resource.Resource) (*resource.Resource, error) {
+	resData, err := res.AsYAML()
+	if err != nil {
+		return nil, err
+	}
+
+	key := fmt.Sprintf("%s/substitute", kustomizev1.GroupVersion.Group)
+
+	if res.GetLabels()[key] == kustomizev1.DisabledValue || res.GetAnnotations()[key] == kustomizev1.DisabledValue {
+		return nil, nil
+	}
+
+	vars := make(map[string]string)
+
+	// load vars from ConfigMaps and Secrets data keys
+	for _, reference := range kustomization.Spec.PostBuild.SubstituteFrom {
+		namespacedName := types.NamespacedName{Namespace: kustomization.Namespace, Name: reference.Name}
+		switch reference.Kind {
+		case "ConfigMap":
+			resource := &corev1.ConfigMap{}
+			if err := kubeClient.Get(ctx, namespacedName, resource); err != nil {
+				return nil, fmt.Errorf("substitute from 'ConfigMap/%s' error: %w", reference.Name, err)
+			}
+			for k, v := range resource.Data {
+				vars[k] = strings.Replace(v, "\n", "", -1)
```

Comment:
Is the idea here to trim trailing white space? If so, wouldn't it be better to use `strings.TrimRight` or `strings.TrimSpace`?

Reply:
Te idea is to remove all the `/n` from the value, to avoid generating an invalid yaml, as the data value can look like this:
```
data:
  some_var: |
      
      bla
      
```